# INSTITUTO FEDERAL GOIÂNO 
## Inteligência Artificial Aplicada a Dados Corporativos
### Trabalho de Conclusão de Trilha de Aprendizagem: I

**Estudante:** Rafael Nakamura Watanabe
**Matrícula:** 2026101806140019  
**Data:** 05 de junho de 2026

### 1. Apresentação do Aluno e Contexto
> Graduado em Zootecnia, mestre e doutor em Genética e Melhoramento Animal (UNESP - FCAV). Atua como Cientista de Dados em empresa de consultoria, trabalhando em projetos para SEFAZ - RS.

> Para este Trabalho de Conclusão de Trilha de Aprendizado, selecionei o contexto de Transparência Pública e Análise de Financiamento Eleitoral. A motivação surgiu de uma dificuldade pessoal: encontrar uma fonte única, confiável e estruturada que permita compreender como os recursos financeiros circulam dentro de uma campanha eleitoral e quais relações podem existir entre fontes de financiamento, partidos políticos, candidatos e setores econômicos que prestam serviços durante o período eleitoral.

> Os dados utilizados foram obtidos a partir das bases públicas disponibilizadas pelo Tribunal Superior Eleitoral (TSE), contemplando receitas e despesas das campanhas presidenciais de 2022. Embora essas informações estejam disponíveis ao cidadão, sua análise exige o cruzamento de múltiplos conjuntos de dados e a interpretação de milhares de registros financeiros, tornando difícil identificar padrões e relações relevantes apenas por meio de consultas manuais.

> Nesse contexto, a Inteligência Artificial e os algoritmos de busca em grafos se apresentam como ferramentas capazes de transformar grandes volumes de dados em conhecimento interpretável. Ao modelar as campanhas eleitorais como uma rede de relacionamentos entre origens de recursos, fontes de financiamento, partidos, candidatos, tipos de despesas e setores econômicos fornecedores, torna-se possível visualizar fluxos financeiros, identificar conexões relevantes e analisar caminhos de influência econômica dentro do ecossistema eleitoral.

> Para este projeto foi utilizada uma abordagem baseada em Grafos e no Algoritmo de Dijkstra. Os dados financeiros foram representados como uma rede direcionada e ponderada, na qual os pesos das conexões refletem a intensidade financeira dos relacionamentos observados. A partir dessa modelagem, foi possível identificar os caminhos de menor custo entre diferentes componentes da rede, interpretados neste trabalho como os fluxos financeiros mais relevantes entre origens de financiamento e setores econômicos envolvidos nas campanhas.

> Embora o projeto represente apenas uma pequena parcela das análises que poderiam ser realizadas com dados eleitorais públicos, ele demonstra como técnicas clássicas de Inteligência Artificial podem contribuir para aumentar a transparência, facilitar a interpretação de informações complexas e apoiar análises mais estruturadas sobre o financiamento de campanhas eleitorais.


### 2. Introdução e Definição do Problema

No contexto da transparência pública e da participação democrática, um dos grandes desafios enfrentados pelos cidadãos não está na falta de dados, mas na dificuldade de transformar informações públicas dispersas em conhecimento útil para apoiar decisões eleitorais. Embora diversos órgãos governamentais disponibilizem bases abertas contendo informações sobre receitas e despesas de campanhas eleitorais, a análise integrada desses dados ainda exige elevado conhecimento técnico e considerável esforço manual por parte do cidadão.

Este trabalho concentra-se especificamente na análise das campanhas presidenciais brasileiras de 2022. Atualmente, informações relacionadas ao financiamento eleitoral encontram-se distribuídas em diferentes conjuntos de dados, tabelas e relatórios disponibilizados pelo Tribunal Superior Eleitoral (TSE). Apesar da elevada transparência institucional, a simples disponibilização dos dados não garante sua compreensão. Para um eleitor comum, torna-se extremamente difícil identificar como os recursos financeiros percorrem diferentes entidades até serem convertidos em ações de campanha.

O processo atual de investigação sobre financiamento eleitoral apresenta três desafios principais:

#### 2.1 Fragmentação das Informações

A análise completa de uma campanha exige o cruzamento de diferentes bases de dados contendo receitas, despesas, partidos políticos, candidatos e fornecedores. Frequentemente, essas informações encontram-se distribuídas em milhares de registros, dificultando a visualização das relações existentes entre as entidades envolvidas.

Por exemplo, um eleitor pode identificar que determinado candidato recebeu recursos oriundos do Fundo Especial de Financiamento de Campanha, mas dificilmente conseguirá rastrear de forma intuitiva como esses recursos foram posteriormente direcionados para diferentes tipos de despesas e setores econômicos.

#### 2.2 Elevada Complexidade de Análise

Mesmo quando os dados estão disponíveis, compreender os fluxos financeiros exige a realização de filtros, agregações e cruzamentos complexos. A simples identificação dos principais caminhos percorridos pelos recursos financeiros entre suas origens e seus destinos finais demanda conhecimento técnico em análise de dados e manipulação de bases relacionais.

Esse processo dificulta a comparação objetiva entre candidatos e limita a capacidade de interpretação dos dados por parte da sociedade.

#### 2.3 Dificuldade de Identificação de Padrões

A análise tradicional baseada em tabelas e planilhas apresenta limitações para revelar relações indiretas entre diferentes elementos do ecossistema eleitoral. Muitas vezes, conexões relevantes entre fontes de financiamento, partidos políticos, candidatos e setores econômicos fornecedores permanecem ocultas em meio a grandes volumes de dados.

Como consequência, torna-se difícil responder perguntas como:

* Quais são os principais caminhos percorridos pelos recursos financeiros dentro de uma campanha?
* Quais setores econômicos concentram maior volume de recursos?
* Existem diferenças significativas entre os fluxos financeiros observados em diferentes candidaturas?
* Quais relações apresentam maior relevância econômica dentro da rede de financiamento eleitoral?

O problema central pode ser descrito como uma dificuldade de representar e interpretar relações complexas presentes em grandes volumes de dados financeiros eleitorais.

Para este trabalho, foram utilizadas bases públicas do Tribunal Superior Eleitoral contendo informações sobre:

* Receitas de campanha;
* Despesas contratadas de campanha;
* Fontes de financiamento eleitoral;
* Partidos políticos;
* Candidatos à Presidência da República;
* Categorias econômicas dos fornecedores envolvidos nas campanhas.

Nesse contexto, propõe-se uma solução baseada em Inteligência Artificial utilizando Grafos e o Algoritmo de Dijkstra. Os dados financeiros são modelados como uma rede direcionada e ponderada, na qual cada nó representa um elemento relevante do ecossistema eleitoral e cada aresta representa uma relação financeira entre esses elementos.

A partir dessa representação, torna-se possível identificar os caminhos financeiramente mais relevantes dentro da rede, comparar estruturas de financiamento entre diferentes candidatos e visualizar de forma intuitiva os principais fluxos de recursos observados nas campanhas eleitorais analisadas. O objetivo deste trabalho não é determinar preferências políticas ou indicar candidatos, mas demonstrar como técnicas clássicas de Inteligência Artificial podem auxiliar na interpretação de dados públicos complexos e contribuir para análises mais transparentes e fundamentadas.



### 3. Metodologia e Escolha da Técnica de IA

Para solucionar o problema de interpretação dos fluxos financeiros presentes nas campanhas eleitorais, a metodologia deste trabalho afasta-se de abordagens preditivas tradicionais de Aprendizado de Máquina. O objetivo não é prever resultados eleitorais, estimar intenções de voto ou realizar classificações automáticas. O foco está na representação, exploração e interpretação de relações existentes em dados públicos já disponíveis.

Nesse contexto, a utilização de modelos estatísticos ou redes neurais não se mostra a alternativa mais adequada, uma vez que o problema central não envolve previsão, mas sim a identificação de conexões e caminhos relevantes entre diferentes entidades do ecossistema eleitoral. O desafio consiste em compreender como recursos financeiros percorrem uma rede composta por fontes de financiamento, partidos políticos, candidatos, tipos de despesas e setores econômicos fornecedores.

Por essa razão, a técnica de Inteligência Artificial selecionada foi a modelagem baseada em Grafos combinada com o Algoritmo de Dijkstra para cálculo de caminhos mínimos. De acordo com Russell e Norvig (2022), problemas que envolvem navegação, conectividade e descoberta de caminhos podem ser representados de forma natural por estruturas de grafos, nas quais entidades são modeladas como nós e seus relacionamentos como arestas.

A escolha se justifica porque os dados eleitorais analisados apresentam uma estrutura inerentemente relacional. Um recurso financeiro não existe de forma isolada; ele percorre diferentes entidades até atingir seu destino final. Dessa forma, representar essas relações como uma rede permite visualizar e analisar conexões que dificilmente seriam percebidas por meio de tabelas tradicionais ou consultas manuais.

Além disso, o Algoritmo de Dijkstra possibilita identificar os caminhos de menor custo entre dois pontos da rede. Neste trabalho, o conceito clássico de distância foi adaptado para representar a intensidade financeira dos relacionamentos observados. Assim, caminhos com maior volume financeiro agregado apresentam menor custo computacional, sendo interpretados como os fluxos economicamente mais relevantes dentro da rede eleitoral.

#### 3.1 Formalização do Modelo em Grafos

Para estruturar a rede de relacionamentos, os elementos presentes nas bases do Tribunal Superior Eleitoral (TSE) foram modelados como nós do grafo:

##### 3.1.1 Conjunto de Nós ($V$)

Os principais tipos de entidades considerados foram:

* Origens de Receita;
* Fontes de Receita;
* Partidos Políticos;
* Candidatos;
* Tipos de Despesa;
* Categorias Econômicas dos Fornecedores (CNAE).

Formalmente:

$$
V = {OrigemReceita, FonteReceita, Partido, Candidato, TipoDespesa, CNAEFornecedor}
$$

##### 3.1.2 Conjunto de Arestas ($E$)

As relações financeiras observadas nos dados foram modeladas como arestas direcionadas:

$$
OrigemReceita \rightarrow FonteReceita
$$

$$
FonteReceita \rightarrow Partido
$$

$$
Partido \rightarrow Candidato
$$

$$
Candidato \rightarrow TipoDespesa
$$

$$
TipoDespesa \rightarrow CNAEFornecedor
$$

Cada aresta representa um fluxo financeiro agregado identificado nas bases de receitas e despesas eleitorais.

##### 3.1.3 Definição dos Pesos

Para permitir a aplicação do Algoritmo de Dijkstra, cada aresta recebeu um peso associado ao volume financeiro movimentado.

Seja $v_{ij}$ o valor financeiro agregado entre os nós $i$ e $j$.

O custo da aresta foi definido como:

$$
w_{ij}=\frac{1}{1+\ln(1+v_{ij})}
$$

Essa transformação possui duas propriedades desejáveis:

* Valores financeiros elevados produzem custos menores;
* Diferenças extremas de escala são suavizadas pela função logarítmica.

Dessa forma, conexões economicamente mais relevantes tornam-se mais atrativas para o algoritmo de busca.

###### 3.1.4 Algoritmo de Dijkstra

Considere um grafo direcionado e ponderado:

$$
G=(V,E)
$$

Dado um nó de origem $s$ e um nó de destino $t$, o Algoritmo de Dijkstra busca o caminho cuja soma dos custos das arestas seja mínima:

$$
C(P)=\sum_{(i,j)\in P} w_{ij}
$$

onde $P$ representa um caminho possível entre os dois nós.

No contexto deste trabalho, isso significa identificar os fluxos financeiros mais relevantes entre diferentes componentes da rede eleitoral.

Por exemplo, um caminho identificado pelo algoritmo pode assumir a seguinte estrutura:

$$
Recursos\ de\ Pessoas\ Físicas
\rightarrow
Outros\ Recursos
\rightarrow
Partido
\rightarrow
Candidato
\rightarrow
Impulsionamento\ de\ Conteúdo
\rightarrow
Setor\ Econômico\ Fornecedor
$$

O custo total desse caminho representa uma medida agregada da intensidade financeira observada ao longo da rede.

#### 3.2 Processo Metodológico

A metodologia empregada foi composta pelas seguintes etapas:

1. Coleta das bases públicas de receitas e despesas eleitorais disponibilizadas pelo TSE;
2. Limpeza, transformação e agregação dos dados financeiros;
3. Construção do grafo direcionado e ponderado;
4. Cálculo dos pesos das arestas com base nos valores financeiros agregados;
5. Aplicação do Algoritmo de Dijkstra para identificação dos caminhos de menor custo;
6. Visualização dos fluxos financeiros por meio de diagramas Sankey;
7. Comparação dos caminhos encontrados para diferentes candidaturas presidenciais.

Essa abordagem permitiu transformar milhares de registros financeiros em uma representação visual e interpretável, possibilitando analisar de forma estruturada as conexões econômicas existentes nas campanhas eleitorais estudadas.



### 4. Implementação da Solução

A implementação foi desenvolvida em Python utilizando as bibliotecas Pandas para manipulação dos dados, NetworkX para construção e análise do grafo e Plotly para visualização dos fluxos financeiros por meio de diagramas Sankey.

Inicialmente, os dados públicos disponibilizados pelo Tribunal Superior Eleitoral (TSE) foram carregados e processados a partir dos arquivos de receitas e despesas das campanhas presidenciais de 2022. As informações foram filtradas para considerar exclusivamente os candidatos à Presidência da República analisados neste trabalho: Luiz Inácio Lula da Silva e Jair Messias Bolsonaro.

#### 4.1 Estruturação dos Dados

Após a etapa de limpeza e transformação, os dados foram agregados em categorias econômicas e financeiras para reduzir a complexidade da rede e tornar a interpretação dos resultados mais intuitiva.

Os principais tipos de nós considerados no grafo foram:

* Origens de Receita;
* Fontes de Receita;
* Partidos Políticos;
* Candidatos;
* Tipos de Despesa;
* Categorias Econômicas dos Fornecedores (CNAE).

As conexões entre esses elementos foram representadas por arestas direcionadas, formando a seguinte estrutura lógica:

$$
OrigemReceita
\rightarrow
FonteReceita
\rightarrow
Partido
\rightarrow
Candidato
\rightarrow
TipoDespesa
\rightarrow
CNAEFornecedor
$$

Cada aresta recebeu como atributo o valor financeiro agregado observado nas bases eleitorais.

#### 4.2 Definição do Problema de Busca

Para aplicação do Algoritmo de Dijkstra, o problema foi formalizado da seguinte forma:

##### 4.2.1 Estado Inicial

O estado inicial corresponde a uma origem de receita presente nas campanhas eleitorais, por exemplo:

$$
OrigemReceita = Recursos\ de\ Pessoas\ Físicas
$$

ou

$$
OrigemReceita = Recursos\ de\ Partido\ Político
$$

##### 4.2.2 Espaço de Estados

O espaço de estados é composto por todos os nós existentes na rede financeira construída a partir dos dados eleitorais.

Formalmente:

$$
S = {Origens,\ Fontes,\ Partidos,\ Candidatos,\ Tipos\ de\ Despesa,\ CNAEs}
$$

##### 4.2.3 Ações (Transições)

Uma ação consiste em percorrer uma aresta válida do grafo.

Por exemplo:

$$
Recursos\ de\ Pessoas\ Físicas
\rightarrow
Outros\ Recursos
$$

ou

$$
Partido
\rightarrow
Candidato
$$

Cada transição representa uma relação financeira observada nos dados.

##### 4.2.4 Custo do Caminho

Para permitir a aplicação do algoritmo, os valores financeiros foram transformados em custos.

Seja $v_{ij}$ o valor financeiro agregado entre dois nós.

O peso da aresta foi definido por:

$$
w_{ij}=\frac{1}{1+\ln(1+v_{ij})}
$$

Essa transformação garante que conexões associadas a maiores volumes financeiros apresentem custos menores, tornando-as mais relevantes durante a busca.

O custo total de um caminho corresponde à soma dos pesos das arestas percorridas.

##### 4.2.5 Objetivo

O objetivo da busca consiste em encontrar o caminho de menor custo entre uma origem de receita e uma categoria econômica fornecedora.

No contexto deste trabalho, um caminho de menor custo representa um fluxo financeiro de maior relevância econômica dentro da rede eleitoral.

#### 4.3 Parametrização do Algoritmo

Foi utilizado o Algoritmo de Dijkstra implementado pela biblioteca NetworkX.

Os principais parâmetros adotados foram:

* Grafo direcionado (`DiGraph`);
* Peso das arestas definido pelo atributo `weight`;
* Origens de busca representadas pelos nós do tipo `origem_receita`;
* Destinos representados pelos nós do tipo `cnae_fornecedor`.

A execução do algoritmo resultou na identificação dos caminhos financeiramente mais relevantes entre as diferentes etapas do fluxo de financiamento eleitoral.

#### 4.4 Principais Funções Implementadas

As principais funções desenvolvidas durante o projeto foram:

* `build_category_edges()`: construção das arestas agregadas da rede;
* `aggregate_edges()`: agregação dos valores financeiros entre pares de nós;
* `build_category_graph()`: construção do grafo direcionado e ponderado;
* `candidate_paths()`: identificação dos caminhos associados a cada candidato;
* `path_cost()`: cálculo do custo total de um caminho;
* `plot_candidate_sankey()`: geração das visualizações Sankey para análise dos fluxos financeiros.

Essas funções permitiram transformar os dados brutos do TSE em uma representação estruturada baseada em grafos, possibilitando a aplicação do Algoritmo de Dijkstra e a interpretação visual dos principais fluxos financeiros observados nas campanhas eleitorais analisadas.



In [22]:
# Importação de bibliotecas
from pathlib import Path
import sys
import requests
sys.path.append(str(Path("..").resolve()))
from src.tse_io import * #as principais funções foram importadas de outra pasta, porém vou apresenta-las aqui também
from src.graph_builder import *
import networkx as nx
import pandas as pd
import plotly.express as px
import plotly.graph_objects as go

In [30]:
# Função utilizada para baixar o arquivo zip do TSE, caso ele não exista localmente
def download_file(url: str, output_path: Path) -> None:
    output_path.parent.mkdir(parents=True, exist_ok=True)

    if output_path.exists():
        print(f"Arquivo já existe: {output_path}")
        return

    print("Baixando arquivo...")
    response = requests.get(url, timeout=120)
    response.raise_for_status()

    output_path.write_bytes(response.content)
    print(f"Arquivo salvo em: {output_path}")

# Função utilizada para ler um arquivo CSV específico de dentro do zip baixado do TSE    
def read_csv_from_zip(
    zip_path: Path,
    file_name: str,
    sep: str = ";",
    encoding: str = "latin1",
) -> pd.DataFrame:
    with zipfile.ZipFile(zip_path, "r") as z:
        if file_name not in z.namelist():
            raise FileNotFoundError(f"Arquivo não encontrado no ZIP: {file_name}")

        with z.open(file_name) as f:
            df = pd.read_csv(
                f,
                sep=sep,
                encoding=encoding,
                dtype=str,
                low_memory=False,
            )

    df["ARQUIVO_ORIGEM"] = file_name
    return df

# Função utilizada para listar os arquivos contidos no zip baixado do TSE    
def list_zip_files(zip_path: Path) -> list[str]:
    with zipfile.ZipFile(zip_path, "r") as z:
        return z.namelist()

# Função utilizada para ler o arquivo Brasil do zip baixado do TSE
def read_brasil_file(zip_path: Path, prefix: str) -> pd.DataFrame:
    file_name = f"{prefix}_BRASIL.csv"
    return read_csv_from_zip(zip_path, file_name)

# Função utilizada para filtrar os candidatos presidenciais do DataFrame (apenas os candidatos com os números 13 e 22 (Lula e Bolsonaro))
def filter_presidential_candidates(
    df: pd.DataFrame,
    candidate_numbers: list[str] | None = None,
) -> pd.DataFrame:
    if candidate_numbers is None:
        candidate_numbers = ["13", "22"]

    out = df.copy()

    if "DS_CARGO" in out.columns:
        out = out[
            normalize_text(out["DS_CARGO"]).eq("PRESIDENTE")
        ].copy()

    if "NR_CANDIDATO" in out.columns:
        out = out[
            out["NR_CANDIDATO"].isin(candidate_numbers)
        ].copy()

    return out

# Função parse para valores monetários, convertendo os valores de string para float, considerando o formato brasileiro
def parse_money(series: pd.Series) -> pd.Series:
    return (
        series.astype(str)
        .str.replace(".", "", regex=False)
        .str.replace(",", ".", regex=False)
        .astype(float)
    )
    
# Função para construção do grafo de categorias, utilizando as funções auxiliares para construir as arestas e agregá-las   
def build_category_graph(receitas_pres, despesas_pres):
    raw_edges = build_category_edges(receitas_pres, despesas_pres)
    edges = aggregate_edges(raw_edges)

    G = nx.DiGraph()

    for _, row in edges.iterrows():
        G.add_edge(
            row["source"],
            row["target"],
            tipo=row["tipo"],
            valor_total=row["valor_total"],
            qtd_registros=row["qtd_registros"],
            weight=row["weight"],
        )

    return G, edges

#Função para limpar os rótulos dos nós, removendo o prefixo de categoria (ex: "categoria::") para uma visualização mais limpa nos gráficos
def clean_label(node):
    return node.split("::", 1)[1] if "::" in node else node

#Função para extrair o subgrafo de um candidato específico, incluindo as arestas de entrada, saída, e as arestas conectadas aos nós vizinhos (upstream e downstream)
def get_candidate_subgraph_edges(edges, candidate_name):
    candidate_node = f"candidato::{candidate_name}"

    incoming = edges[
        edges["target"].eq(candidate_node)
    ].copy()

    outgoing = edges[
        edges["source"].eq(candidate_node)
    ].copy()

    previous_nodes = incoming["source"].unique()

    upstream = edges[
        edges["target"].isin(previous_nodes)
    ].copy()

    next_nodes = outgoing["target"].unique()

    downstream = edges[
        edges["source"].isin(next_nodes)
    ].copy()

    candidate_edges = pd.concat(
        [upstream, incoming, outgoing, downstream],
        ignore_index=True
    ).drop_duplicates()

    return candidate_edges

#Função para plotar o gráfico de Sankey para um candidato específico, utilizando o subgrafo extraído pela função anterior, e considerando apenas as top_n arestas mais relevantes
def plot_candidate_sankey(edges, candidate_name, top_n=40):
    df = get_candidate_subgraph_edges(edges, candidate_name)

    df = df.sort_values("valor_total", ascending=False).head(top_n)

    labels = pd.Index(
        pd.concat([df["source"], df["target"]]).unique()
    )

    node_id = {label: i for i, label in enumerate(labels)}

    fig = go.Figure(
        data=[
            go.Sankey(
                node=dict(
                    label=[clean_label(x) for x in labels],
                    pad=15,
                    thickness=15,
                ),
                link=dict(
                    source=df["source"].map(node_id),
                    target=df["target"].map(node_id),
                    value=df["valor_total"],
                    customdata=df["tipo"],
                    hovertemplate=(
                        "Origem: %{source.label}<br>"
                        "Destino: %{target.label}<br>"
                        "Valor: R$ %{value:,.2f}<br>"
                        "Tipo: %{customdata}<extra></extra>"
                    ),
                ),
            )
        ]
    )

    fig.update_layout(
        title_text=f"Fluxo agregado da campanha: {candidate_name}",
        font_size=10,
        height=700,
    )

    return fig

#Função para calcular o custo total de um caminho no grafo, somando os pesos das arestas ao longo do caminho
def path_cost(G, path):
    cost = 0

    for i in range(len(path) - 1):
        cost += G[path[i]][path[i + 1]]["weight"]

    return cost

#Função para encontrar os caminhos mais curtos (com menor custo) entre as origens de receita e os destinos de fornecedores, passando por um candidato específico
def candidate_paths(
    G,
    candidate_name,
):
    candidate_node = f"candidato::{candidate_name}"

    origens = [
        n for n in G.nodes
        if n.startswith("origem_receita::")
    ]

    destinos = [
        n for n in G.nodes
        if n.startswith("cnae_fornecedor::")
    ]

    rows = []

    for origem in origens:
        for destino in destinos:

            try:
                path = nx.shortest_path( #Dijkstra
                    G,
                    source=origem,
                    target=destino,
                    weight="weight",
                )

                if candidate_node not in path:
                    continue

                rows.append({
                    "origem": origem,
                    "destino": destino,
                    "candidato": candidate_name,
                    "custo": path_cost(G, path),
                    "caminho": " -> ".join(path),
                    "num_passos": len(path) - 1,
                })

            except nx.NetworkXNoPath:
                pass

    return pd.DataFrame(rows)

#Função para descrever um caminho específico no grafo, detalhando as arestas e seus atributos ao longo do caminho
def describe_path(G, path):
    rows = []

    for i in range(len(path) - 1):
        source = path[i]
        target = path[i + 1]
        edge = G[source][target]

        rows.append({
            "step": i + 1,
            "source": source,
            "target": target,
            "tipo": edge["tipo"],
            "valor_total": edge["valor_total"],
            "qtd_registros": edge["qtd_registros"],
            "weight": edge["weight"],
        })

    return pd.DataFrame(rows)

In [ ]:
TSE_URL = (
    "https://cdn.tse.jus.br/estatistica/sead/odsele/prestacao_contas/"
    "prestacao_de_contas_eleitorais_candidatos_2022.zip"
)

zip_path = Path("data/raw/prestacao_de_contas_eleitorais_candidatos_2022.zip")
download_file(TSE_URL, zip_path)

Baixando arquivo...
Arquivo salvo em: data/raw/prestacao_de_contas_eleitorais_candidatos_2022.zip


In [ ]:
# Printando os arquivos contidos no zip baixado do TSE
zip_path = Path("/home/rafaelnw/aula_if_go/t1_trabalho_conclusao/data/raw/prestacao_de_contas_eleitorais_candidatos_2022.zip")
files = list_zip_files(zip_path)
files[:10]

['despesas_pagas_candidatos_2022_BRASIL.csv',
 'receitas_candidatos_2022_RR.csv',
 'despesas_pagas_candidatos_2022_TO.csv',
 'despesas_pagas_candidatos_2022_SP.csv',
 'receitas_candidatos_2022_BA.csv',
 'receitas_candidatos_2022_SE.csv',
 'receitas_candidatos_2022_MT.csv',
 'despesas_pagas_candidatos_2022_MA.csv',
 'receitas_candidatos_2022_SP.csv',
 'receitas_candidatos_2022_RN.csv']

In [5]:
# Lendo os arquivos de receitas e despesas contratadas do zip, apenas dos arquivos do Brasil (excluindo estados)
receitas = read_brasil_file(
    zip_path,
    "receitas_candidatos_2022"
)

despesas_contratadas = read_brasil_file(
    zip_path,
    "despesas_contratadas_candidatos_2022"
)

In [ ]:
#Filtrando os candidatos presidenciais (Lula e Bolsonaro) dos DataFrames de receitas e despesas contratadas
receitas_pres = filter_presidential_candidates(receitas)
despesas_pres = filter_presidential_candidates(despesas_contratadas)

### 5. Resultados e Comparações

In [ ]:
#Aplicando função de parse para valores monetários, convertendo os valores de string para float, considerando o formato brasileiro, nas colunas de valores de receitas e despesas contratadas
receitas_pres["VR_RECEITA_NUM"] = parse_money(receitas_pres["VR_RECEITA"])
despesas_pres["VR_DESPESA_CONTRATADA_NUM"] = parse_money(
    despesas_pres["VR_DESPESA_CONTRATADA"]
)

In [ ]:
# Plotando gráfico de barras comparando as receitas e despesas totais dos candidatos presidenciais
receita_total = (
    receitas_pres
    .groupby("NM_CANDIDATO", as_index=False)
    .agg(valor=("VR_RECEITA_NUM", "sum"))
)

receita_total["tipo"] = "Receita"

despesa_total = (
    despesas_pres
    .groupby("NM_CANDIDATO", as_index=False)
    .agg(valor=("VR_DESPESA_CONTRATADA_NUM", "sum"))
)

despesa_total["tipo"] = "Despesa"

financeiro_total = pd.concat(
    [receita_total, despesa_total],
    ignore_index=True
)

fig = px.bar(
    financeiro_total,
    x="NM_CANDIDATO",
    y="valor",
    color="tipo",
    barmode="group",
    title="Receitas e despesas totais por candidato",
    labels={
        "NM_CANDIDATO": "Candidato",
        "valor": "Valor total (R$)",
        "tipo": "Tipo"
    },
    text_auto=".2s"
)

fig.show()

Bolsonaro apresentou receitas e despesas totais menores se comparado com Lula, com maior diferença nos valores de despesas (89M x 130M).

In [20]:
receita_origem_fonte = (
    receitas_pres
    .groupby(
        ["NM_CANDIDATO", "DS_ORIGEM_RECEITA", "DS_FONTE_RECEITA"],
        as_index=False
    )
    .agg(valor=("VR_RECEITA_NUM", "sum"))
)

receita_origem_fonte["categoria"] = (
    receita_origem_fonte["DS_ORIGEM_RECEITA"]
    + " | "
    + receita_origem_fonte["DS_FONTE_RECEITA"]
)

fig = px.bar(
    receita_origem_fonte,
    x="valor",
    y="NM_CANDIDATO",
    color="categoria",
    orientation="h",
    title="Composição das receitas por origem e fonte",
    labels={
        "valor": "Valor total (R$)",
        "NM_CANDIDATO": "Candidato",
        "categoria": "Origem | Fonte"
    },
    text_auto=".2s"
)

fig.show()

Maior fonte de receita para Lula foi com Recursos de partido político | FUNDO ESPECIAL (130M).

Maior fonte de receita para Bolsonaro foi com Recursos de pessoas físicas | OUTROS RECURSOS.

In [21]:
despesa_tipo = (
    despesas_pres
    .groupby(
        ["NM_CANDIDATO", "DS_ORIGEM_DESPESA"],
        as_index=False
    )
    .agg(valor=("VR_DESPESA_CONTRATADA_NUM", "sum"))
)

top_despesa_tipo = (
    despesa_tipo
    .sort_values("valor", ascending=False)
    .groupby("NM_CANDIDATO")
    .head(8)
)

fig = px.bar(
    top_despesa_tipo,
    x="valor",
    y="DS_ORIGEM_DESPESA",
    color="NM_CANDIDATO",
    orientation="h",
    title="Principais tipos de despesa por candidato",
    labels={
        "valor": "Valor total (R$)",
        "DS_ORIGEM_DESPESA": "Tipo de despesa",
        "NM_CANDIDATO": "Candidato"
    },
    text_auto=".2s"
)

fig.update_layout(
    yaxis={"categoryorder": "total ascending"}
)

fig.show()

Maior despesa de Lula foi com Produção de programas de rádio, televisão ou vídeo (38M).

Maior depesa de Bolsonaro foi com Despesas com Impulsionamento de Conteúdos (33M).

In [ ]:
# Construindo o grafo de categorias, utilizando as funções auxiliares para construir as arestas e agregá-las
G_cat, edges_cat = build_category_graph(
    receitas_pres=receitas_pres,
    despesas_pres=despesas_pres,
)

G_cat.number_of_nodes(), G_cat.number_of_edges()

(144, 224)

O grafo final possui 144 nós e 224 arestas. Essa agregação reduz a complexidade dos dados originais e permite interpretar os principais fluxos financeiros entre receitas, partidos, candidatos, despesas e setores fornecedores.

In [24]:
fig_lula = plot_candidate_sankey(
    edges_cat,
    "LUIZ INÁCIO LULA DA SILVA",
    top_n=40
)

fig_lula.show()

In [25]:
fig_bolsonaro = plot_candidate_sankey(
    edges_cat,
    "JAIR MESSIAS BOLSONARO",
    top_n=40
)

fig_bolsonaro.show()

Os diagramas Sankey mostram diferenças na estrutura dos fluxos financeiros. A campanha de Lula apresenta maior concentração em recursos associados ao Fundo Especial e ao PT. Já a campanha de Bolsonaro apresenta maior presença de Outros Recursos e maior diversidade visual de conexões com tipos de despesas e setores fornecedores.

### Aplicação de Dijkstra

In [28]:
paths_bolsonaro = candidate_paths(
    G_cat,
    "JAIR MESSIAS BOLSONARO"
)

paths_lula = candidate_paths(
    G_cat,
    "LUIZ INÁCIO LULA DA SILVA"
)

comparacao = pd.concat(
    [
        paths_lula,
        paths_bolsonaro,
    ],
    ignore_index=True
)

comparacao.groupby(
    "candidato",
    as_index=False
).agg(
    menor_custo=("custo", "min"),
    custo_medio=("custo", "mean"),
    qtd_caminhos=("custo", "count"),
)

,candidato,menor_custo,custo_medio,qtd_caminhos
0,JAIR MESSIAS BOLSONARO,0.262204,0.324637,285
1,LUIZ INÁCIO LULA DA SILVA,0.260865,0.325754,195


Com o grafo construído, aplicou-se o Algoritmo de Dijkstra para identificar os caminhos de menor custo entre origens de receita e categorias econômicas fornecedoras. Como o custo das arestas foi definido de forma inversamente proporcional ao valor financeiro agregado, os caminhos mínimos representam os fluxos economicamente mais relevantes.

Os custos mínimos e médios dos caminhos são bastante semelhantes entre os dois candidatos, indicando que os fluxos financeiros mais intensos apresentam força agregada comparável. 

A principal diferença observada está na quantidade de caminhos identificados: a rede associada a Jair Bolsonaro apresentou maior número de conexões possíveis entre origens de receita e categorias econômicas fornecedoras, sugerindo maior diversidade relacional no grafo.

### Exemplo de caminho mínimo identificado

A tabela abaixo apresenta o caminho de menor custo encontrado pelo algoritmo de Dijkstra em toda a rede. Como os custos das arestas foram definidos de forma inversamente proporcional ao valor financeiro agregado, esse caminho pode ser interpretado como um dos fluxos financeiramente mais relevantes do grafo.

In [31]:
melhor_caminho = comparacao.sort_values("custo").iloc[0]

source = melhor_caminho["origem"]
target = melhor_caminho["destino"]

path = nx.shortest_path(
    G_cat,
    source=source,
    target=target,
    weight="weight",
)

describe_path(G_cat, path)

,step,source,target,tipo,valor_total,qtd_registros,weight
0,1,origem_receita::Recursos de partido político,fonte_receita::FUNDO ESPECIAL,origem_fonte,1.270427e+08,310,0.050865
1,2,fonte_receita::FUNDO ESPECIAL,partido::PT,fonte_partido,1.251190e+08,262,0.050904
2,3,partido::PT,candidato::LUIZ INÁCIO LULA DA SILVA,partido_candidato,1.355393e+08,533,0.050698
3,4,candidato::LUIZ INÁCIO LULA DA SILVA,"tipo_despesa::Produção de programas de rádio, ...",candidato_tipo_despesa,3.789900e+07,2,0.054199
4,5,"tipo_despesa::Produção de programas de rádio, ...",cnae_fornecedor::Agências de publicidade,tipo_despesa_cnae_fornecedor,3.789900e+07,2,0.054199


Entre todas as combinações possíveis de origem de receita e categoria econômica fornecedora, o caminho de menor custo identificado passa pela campanha de Lula, conectando recursos de partido político ao Fundo Especial, ao PT, ao candidato, à despesa com produção audiovisual e ao setor de agências de publicidade.

### 6. Perspectiva de Evolução do Trabalho

Embora a modelagem baseada em Grafos e o Algoritmo de Dijkstra tenha se mostrado adequada para representar e analisar fluxos financeiros eleitorais, uma possível evolução deste trabalho seria a incorporação de técnicas de Deep Learning capazes de identificar padrões complexos de financiamento e comportamento financeiro das campanhas eleitorais.

Enquanto a abordagem atual possui caráter predominantemente explicativo e exploratório, uma Rede Neural Profunda poderia agregar capacidades preditivas e de descoberta automática de padrões, permitindo analisar grandes volumes de dados históricos de eleições anteriores e identificar estruturas financeiras semelhantes entre diferentes campanhas.

#### 6.1 Organização e Engenharia dos Dados de Entrada (X)

Os dados de entrada continuariam sendo obtidos a partir das bases públicas do Tribunal Superior Eleitoral (TSE), porém com uma abrangência maior, contemplando múltiplos ciclos eleitorais.

As principais variáveis coletadas poderiam incluir:

* Valores de receitas recebidas;
* Origens das receitas;
* Fontes de financiamento;
* Partidos políticos;
* Cargo disputado;
* Tipo de despesa realizada;
* Categoria econômica dos fornecedores (CNAE);
* Quantidade de fornecedores distintos;
* Quantidade de doadores distintos;
* Distribuição percentual dos gastos por categoria;
* Informações temporais da campanha.

Como a maior parte dessas informações possui natureza tabular, seria necessário realizar uma etapa de engenharia de atributos.

As variáveis numéricas seriam normalizadas utilizando técnicas como Min-Max Scaling ou Standard Scaling:

$$
x_{norm}=\frac{x-\mu}{\sigma}
$$

As variáveis categóricas, como partido político, origem de receita e CNAE, poderiam ser representadas por meio de embeddings aprendidos pela própria rede neural, permitindo capturar relações semânticas entre categorias.

O vetor final de entrada seria composto pela concatenação das variáveis numéricas normalizadas e dos embeddings das variáveis categóricas.

#### 6.2 Representação e Arquitetura Teórica da Rede

Como os dados possuem estrutura predominantemente tabular, uma arquitetura baseada em camadas densas (*Fully Connected Neural Networks*) seria a mais adequada.

Uma arquitetura hipotética poderia ser composta por:

* Camada de Entrada;
* Camada de Embedding para atributos categóricos;
* Camada Densa com 128 neurônios;
* Camada Densa com 64 neurônios;
* Camada Densa com 32 neurônios;
* Camada de Saída.

A função de ativação utilizada nas camadas ocultas seria a ReLU:

$$
ReLU(x)=max(0,x)
$$

A utilização da ReLU permite à rede capturar relações não lineares entre as variáveis financeiras, algo comum em sistemas complexos de financiamento eleitoral.

Para reduzir o risco de sobreajuste (*overfitting*), poderiam ser adicionadas camadas de Dropout entre as camadas densas.

#### 6.3 Estrutura da Saída Esperada (Y)

A estrutura da saída dependeria do objetivo de negócio definido para a solução.

Uma possibilidade seria formular o problema como uma classificação multiclasses, na qual a rede deveria identificar o perfil predominante de uma campanha eleitoral.

Exemplos de classes:

* Predominância de financiamento por pessoas físicas;
* Predominância de financiamento partidário;
* Predominância de Fundo Especial;
* Perfil financeiro híbrido.

Nesse cenário, a camada de saída utilizaria a função Softmax:

$$
Softmax(z_i)=\frac{e^{z_i}}{\sum_{j=1}^{K} e^{z_j}}
$$

A saída da rede corresponderia a uma distribuição de probabilidades para cada perfil de campanha.

Como função de perda, seria utilizada a Categorical Cross Entropy:

$$
L=-\sum_{i=1}^{K} y_i\log(\hat{y}_i)
$$

onde:

* $y_i$ representa a classe real;
* $\hat{y}_i$ representa a probabilidade prevista pela rede.

O treinamento seria realizado utilizando o algoritmo Adam, amplamente empregado em problemas de classificação tabular devido à sua estabilidade e eficiência computacional.

### 7. Conclusões

Este trabalho demonstrou como técnicas clássicas de Inteligência Artificial podem ser utilizadas para transformar grandes volumes de dados públicos em informações mais acessíveis e interpretáveis. A partir dos dados de receitas e despesas das campanhas presidenciais de 2022 disponibilizados pelo Tribunal Superior Eleitoral (TSE), foi possível modelar os fluxos financeiros por meio de grafos direcionados e ponderados, permitindo visualizar as principais relações entre origens de recursos, partidos políticos, candidatos, tipos de despesas e setores econômicos fornecedores.

A aplicação do Algoritmo de Dijkstra permitiu identificar os caminhos financeiramente mais relevantes dentro da rede construída, enquanto os diagramas Sankey auxiliaram na interpretação visual desses fluxos. Os resultados mostraram que as campanhas analisadas apresentam estruturas financeiras com características distintas, mas com custos mínimos e médios bastante semelhantes segundo a função de custo adotada. Além disso, observou-se uma diferença na quantidade de caminhos identificados para cada candidato, sugerindo diferentes níveis de diversidade relacional na rede.

A solução proposta mostrou-se viável para aplicações relacionadas à transparência pública e análise exploratória de dados eleitorais, apresentando como principal vantagem a sua capacidade de explicação. Diferentemente de modelos mais complexos e menos interpretáveis, a abordagem baseada em grafos permite reconstruir e justificar cada caminho encontrado pelo algoritmo, facilitando auditorias e análises independentes. Como evolução futura, o projeto poderia ser expandido para múltiplos ciclos eleitorais e incorporar técnicas mais avançadas de análise de redes e aprendizado de máquina.



### 8. Referências Bibliográficas

RUSSELL, Stuart; NORVIG, Peter. *Artificial Intelligence: A Modern Approach*. 4. ed. Pearson, 2022.

CORMEN, Thomas H.; LEISERSON, Charles E.; RIVEST, Ronald L.; STEIN, Clifford. *Introduction to Algorithms*. 4. ed. MIT Press, 2022.

NETWORKX DEVELOPERS. *NetworkX Documentation*. Disponível em: https://networkx.org/documentation/stable/. Acesso em: 04 jun. 2026.

PANDAS DEVELOPMENT TEAM. *Pandas Documentation*. Disponível em: https://pandas.pydata.org/docs/. Acesso em: 04 jun. 2026.

PLOTLY TECHNOLOGIES INC. *Plotly Python Documentation*. Disponível em: https://plotly.com/python/. Acesso em: 04 jun. 2026.

TRIBUNAL SUPERIOR ELEITORAL (TSE). *Dados Abertos - Prestação de Contas Eleitorais 2022*. Disponível em: https://dadosabertos.tse.jus.br/. Acesso em: 04 jun. 2026.

\\begin{titlepage}
    \\centering
    
    % 1. Logotipo do IF Goiano Local
    \\includegraphics[width=0.15\\textwidth]{logo_urutai.png} \\\\[0.2cm]
    \\vspace{1cm}
    
    % 2. Cabeçalho Institucional
    {\\small \\textbf{INSTITUTO FEDERAL GOIANO} \\\\}
    {\\small \\textbf{CAMPUS URUTAÍ} \\\\}
    {\\footnotesize NÚCLEO DE INFORMÁTICA \\\\}
    {\\footnotesize ESPECIALIZAÇÃO EM INTELIGÊNCIA ARTIFICIAL APLICADA A DADOS CORPORATIVOS \\\\}
    \\vspace{2cm}
    
    % 3. Nome do Autor
    {\\Large \\MakeUppercase{Rafael Nakamura Watanabe} \\par}
    \\vspace{4.5cm}
    
    % 4. Título do Trabalho (Negrito e Centralizado)
    {\\Large \\textbf{\\MakeUppercase{ANÁLISE DE FLUXOS FINANCEIROS EM CAMPANHAS ELEITORAIS UTILIZANDO GRAFOS E ALGORITMO DE DIJKSTRA}} \\par}
    
    \\vfill
    
    % 5. Rodapé (Local e Data)
    {\\large Urutaí, \\today \\par}

\\end{titlepage}
\\newpage



\\begin{titlepage}
    \\centering
    
    % 1. Cabeçalho Institucional Simples
    {\\small \\textbf{INSTITUTO FEDERAL GOIANO} \\\\}
    {\\small \\textbf{CAMPUS URUTAÍ} \\\\}
    {\\footnotesize NÚCLEO DE INFORMÁTICA \\\\}
    {\\footnotesize INTELIGÊNCIA ARTIFICIAL APLICADA A DADOS CORPORATIVOS \\\\}
    \\vspace{2.5cm}
    
    % 2. Nome do Autor
    {\\Large RAFAEL NAKAMURA WATANABE \\par}
    \\vspace{3.5cm}
    
    % 3. Título do Trabalho
    {\\Large \\textbf{ANÁLISE DE FLUXOS FINANCEIROS EM CAMPANHAS ELEITORAIS UTILIZANDO GRAFOS E ALGORITMO DE DIJKSTRA} \\par}
    \\vspace{3cm}
    
    % 4. Bloco de Texto de Apresentação (Recuado à Direita)
    \\hspace{0.45\\textwidth} % Empurra o bloco para a metade direita da folha
    \\begin{minipage}{0.5\\textwidth}
        \\small
        \\singlespacing % Mantém espaçamento simples na nota, se necessário
        Relatório Técnico apresentado ao Núcleo de Informática, curso de Inteligência Artificial Aplicada a Dados Corporativos, do Instituto Federal Goiano, como parte das exigências para aprovação na Trilha de Aprendizado I.
    \\end{minipage}
    \\vspace{2cm}
    
    % 5. Orientadores (Alinhados ao centro ou à esquerda se preferir)
    {\\small Orientador(a): \\\\ <quando houver> \\par}
    \\vspace{0.4cm}
    {\\small Co-orientador(a): \\\\ <quando houver> \\par}
    
    \\vfill
    
    % 6. Rodapé com a sua Data Dinâmica
    {\\large Urutaí, \\today \\par}

\\end{titlepage}
\\newpage


In [1]:
%%capture
import re

def salvar_apendice_limpo(texto_procurado, nome_arquivo_final):
    """
    Busca no histórico do notebook a célula que contém o 'texto_procurado',
    remove os comandos técnicos e salva o arquivo .py perfeitamente limpo.
    """
    # In é a lista nativa do Jupyter que guarda o histórico de todas as células executadas
    for codigo_celula in In:
        if texto_procurado in codigo_celula:
            linhas = codigo_celula.split('\n')
            linhas_limpas = []

            for linha in linhas:
                # Remove linhas de comandos mágicos do Colab para o arquivo ficar puro
                if linha.strip().startswith('%') or 'salvar_apendice_limpo' in linha:
                    continue
                linhas_limpas.append(linha)

            # Salva o arquivo final estruturado
            conteudo_final = '\n'.join(linhas_limpas).strip()
            with open(nome_arquivo_final, 'w', encoding='utf-8') as f:
                f.write(conteudo_final)
            print(f"✅ Arquivo {nome_arquivo_final} gerado com sucesso!")
            return
    print(f"❌ Célula com o termo '{texto_procurado}' não foi encontrada no histórico.")

# =====================================================================
# INDIQUE UMA FRASE ÚNICA DE CADA CÉLULA E O NOME DO ARQUIVO DESEJADO
# =====================================================================
salvar_apendice_limpo("def build_category_graph", "graph_builder_functions.py")
salvar_apendice_limpo("def candidate_paths", "dijkstra_search_functions.py")


\\newpage
\\makeatletter
\\renewcommand{\\@chapapp}{Apêndice}
\\makeatother
\\appendix

\\section{Funções Implementadas para Construção de Grafos}
\\label{apendice:grafo}

O bloco de código a seguir apresenta a implementação em Python das funções utilizadas para construir o grafo direcionado e ponderado a partir dos dados de receitas e despesas eleitorais.

\\lstinputlisting[language=Python]{graph_builder_functions.py}

\\clearpage
\\section{Funções Implementadas para Busca com Algoritmo de Dijkstra}
\\label{apendice:dijkstra}

O bloco de código a seguir apresenta a implementação em Python das funções utilizadas para identificar os caminhos de menor custo entre as diferentes componentes da rede eleitoral.

\\lstinputlisting[language=Python]{dijkstra_search_functions.py}

\\clearpage
\\section{Especificações de Hardware e Requisitos de Software}
\\label{apendice:requisitos}

Este apêndice detalha as versões exatas de todas as dependências do ecossistema Python e dos pacotes de distribuição do LaTeX (XeTeX) utilizados para a computação e análise dos fluxos financeiros eleitorais.

\\verbatiminput{requisitos_sistema.txt}

